File: javascript/extractor/src/com/semmle/js/ast/StaticInitializer.java
package com.semmle.js.ast;

/**
 * A static initializer block in a class. E.g. ```TypeScript class Foo { static
 * bar : number; static { Foo.bar = 42; } }
 */
public class StaticInitializer extends MemberDefinition<BlockStatement> {
  public StaticInitializer(SourceLocation loc, BlockStatement body) {
    super("StaticInitializer", loc, DeclarationFlags.static_, null, body);
  }

  @Override
  public boolean isConcrete() {
    return false;
  }

  @Override
  public <C, R> R accept(Visitor<C, R> v, C c) {
    return v.visit(this, c);
  }
}

File: javascript/ql/test/library-tests/Classes/staticInitializer.js
class MyClass {
  static x = 1;
  constructor() {
    this.y = 2;
  }
  static {
    MyClass.z = 3;
  }
  foo() {
    this.t = 4;
  }
  static bar() {
    this.u = 5;
  }
  static {
    this.v = 6;
  }
}